### DISCUSSION TITLE
Semantic Line Breaks

### DISCUSSION THREAD

**Discussion by rooterkyberian** [prettier/prettier#14828] [Ideas]
How about enforcing Semantic Line Breaks as described by https://bobheadxi.dev/semantic-line-breaks/ with prettier?

Author of that already prepared an implementation based on remark, but it has multiple issues, and most importantly - only supports Markdown.
Seems like prettier would be a great choice to implement this.

Maybe it could be done through [`--prose-wrap` option](https://prettier.io/docs/en/options.html#prose-wrap)? I.e. by adding `--prose-wrap sentence`? The default behavior could be changed to we wrap by both sentence and linewidth (in that order).

**Comment by kachkaev** [comment]
Some thoughts on this: https://github.com/prettier/prettier/issues/9810#issuecomment-1002696041

TLDR: Detecting an end of the sentence reliably is quite hard, so I would say that this is out of Prettier’s scope. However, if semantic line breaks are included in other linters like Markdownlint or ESLint, it is possible to run them alongside Prettier. AFAIU, Prettier won’t remove line breaks if another too has added them in markdown prose or JS comments.

I’m using semantic line breaks too, so would be interested in seeing links to working tools that supplement Prettier!

**Reply by js-choi** [reply]
As you said, reliably detecting the ends of sentences is difficult. But there is a solution built into the JavaScript language.

You could try using the standard [Intl.Segmenter](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/Segmenter) API with `{ granularity: "sentence" }`. This is a built-in API that is Baseline Newly Available in browsers, so it wouldn’t make Prettify’s package huge.

Of course, the definition of a sentence is locale dependent, so Prettify would need to be configurable with locale(s) like `en`.

In addition, if Prettify used Intl.Segmenter to break sentences into words (which also would isolate punctuation marks), this would allow Prettify to robustly detect phase-delimiting punctuation more easily. Furthermore, Prettier's line wrapping would become more robust with languages that do not use spaces between their words at all, such as Chinese.

(Prettify’s inability to preserve my semantic line breaks in my Markdown and HTML has made me averse to using Prettify. But if Prettify enforced semantic line breaks, then Prettify would become much more attractive to me.)

**Comment by rooterkyberian** [comment]
Well, so far I'm using the aforementioned https://github.com/bobheadxi/readable which is based on regexp for sentence detection. 
It does "good enough" job for English docs.

Like you said, a true comprehensive solution doesn't seem to exists. I have worked in the past with sentence tokenization and for English there were tools that worked well for me, but there were all pretty "heavy" (such as nltk punkt tokenizers) and language specific.

Seems most projects are using English for documentation it seems like something that works decently well for it would be "good enough".
I will see how well the `readable` regexp will fare after use it more.

**Comment by dmyersturnbull** [comment]
I've been doing this for a decade and didn't know it had a name! Nice to see others doing the same thing.
